services/src/DFUService/DFUService.cpp

```diff
@@ -0,0 +1,402 @@
+        /* Reject writes that modify read-only bits */
+        tr_debug("dfu_ctrl write request: rejected (read-only)");
+        params->authorizationReply = (GattAuthCallbackReply_t) AUTH_CALLBACK_REPLY_ATTERR_APP_READONLY;
+        return;
+    }
+
+    if(_ctrl_req_cb) {
+        /* Forward request to the application */
+        params->authorizationReply = _ctrl_req_cb(change);
+        tr_debug("dfu_ctrl write request: accepted (by application)");
+    } else {
+        /* If no application handler, accept by default */
+        tr_debug("dfu_ctrl write request: accepted");
+        params->authorizationReply = AUTH_CALLBACK_REPLY_SUCCESS;
+    }
+}
+
+void DFUService::on_dfu_ctrl_written(uint8_t new_ctrl) {
+    tr_debug("dfu_ctrl written: %d", new_ctrl);
+    mbed::ScopedLock<PlatformMutex> lock(_mutex);
+    ControlChange change(*this, new_ctrl);
+    /* Call application handler for control updates, if available */
+    if(_ctrl_update_cb) {
+        _ctrl_update_cb(change);
+    }
+
+    if(change.get_changed_bits() & DFU_CTRL_ENABLE_BIT) {
+        tr_debug("dfu mode %s", (change.value() & DFU_CTRL_ENABLE_BIT) ? "enabled" : "aborted");
+
+        if(change.value() & DFU_CTRL_ENABLE_BIT) {
+            /* If DFU is being enabled, clear the currently-selected update slot */
+            _queue.call(mbed::callback(this, &DFUService::init_selected_slot));
+        }
+    }
+
+    if(change.get_changed_bits() & DFU_CTRL_DELTA_MODE_EN_BIT) {
+        tr_debug("delta mode %s", (change.value() & DFU_CTRL_DELTA_MODE_EN_BIT) ? "enabled" : "disabled");
+    }
+
+    if(change.get_changed_bits() & DFU_CTRL_COMMIT_BIT) {
+        tr_debug("dfu commit");
+    }
+
+    _dfu_control = new_ctrl;
+}
+
+void DFUService::init_selected_slot(void) {
+    mbed::ScopedLock<PlatformMutex> lock(_mutex); // TODO mutex lock necessary here?
+    tr_debug("initializing slot %d", _selected_slot);
+    mbed::BlockDevice* slot = _slot_bds[_selected_slot];
+    slot->init();
+    slot->erase(0, slot->size());
+    // Send a neutral notification of the status characteristic to tell the client we're ready
+    set_status(DFU_STATE_IDLE);
+}
+
+
+void DFUService::process_buffer(void) {
+
+    /* TODO Rework the whole writing buffered data stuff */
+
+    mbed::BlockDevice* slot = _slot_bds[_selected_slot];
+    /* Attempt to write as many multiples of program size as possible at once */
+
+    /**
+     * TODO is there a better way to do this? Without dynamic allocation, we would
+     * have to program byte-by-byte. This would likely be a significant hit in speed.
+     */
+    bd_size_t write_size = (_bin_stream_buf.size() / slot->get_program_size()) * slot->get_program_size();
+    tr_debug("processing buffer: %lu => %lu",
+            _bin_stream_buf.size(),
+            write_size);
+    if(write_size == 0) {
+        /* Skip 0-length writes */
+        _scheduled_write = 0;
+        return;
+    }
+    uint8_t *temp_buf = new uint8_t[write_size];
```

Comment:
I'm thinking of making a sort of BlockDevice decorator class with a built-in buffer (`BufferedBlockDevice` is already taken :confounded:). This way, the data can be moved from the BLE stack buffers into the BlockDevice buffer and we can eliminate the need to allocate a separate buffer when writing the data.

@pan- The major decision I'm facing with this is: writing to flash can take a long time -- possibly long enough to cause timeouts for BLE. I want to come up with a way that allows the user to decide whether to process the buffered flash writes in separate thread or in the same context as the BLE (for baremetal users).

In my testing, I tried to process flash writes in the BLE thread but it seemed to cause disconnections/timeouts. Though maybe this was related to other issues, as realistically flash writes shouldn't take _that_ long, probably less than the number of milliseconds required to even miss a connection event.

My idea for enabling both the multi-threaded and baremetal use cases is (assuming event queues are available in baremetal) to let the user provide the event queue on which to process flash writes. This way, they can pass in the existing event queue used for BLE events as well _or_ a separate EventQueue that is dispatched on another thread.

Perhaps we also eliminate writing multiples of the BlockDevice's program size and simply queue writes of the program size. The drawback here is it will take more calls to empty the flash buffer (ie: longer overall write times). However, each write event will be shorter and prevent the flash writes from blocking the BLE stack for too long.

We could potentially make this configurable as well if there's any real performance difference between the two methods.